2022/week11/tp11_advanced.tex

```diff
@@ -0,0 +1,108 @@
+\input{../header.tex}
+\usepackage{array}
+\newcolumntype{C}[1]{>{\centering\let\newline\\\arraybackslash\hspace{0pt}}m{#1}}
+
+\begin{document}
+% Change the following values to true to show the solutions or/and the hints
+\ShowSolutionfalse
+\ShowConseiltrue
+\titre
+\cours{Algorithmes de recherche - Exercices avancés}
+
+
+\begin{comment}
+\begin{enumerate}
+    \item Recherche séquentielle
+    \item Recherche binaire
+    \item Arbres de recherche binaire\\
+\end{enumerate}
+\end{comment}
+
+Le but de cette séance est de se familiariser avec les algorithmes de recherche. Dans la série d'exercices, nous manipulerons des listes et collections en Java et Python. Nous reviendrons sur la notion de récursivité et découvrirons les arbres de recherche. Au terme de cette séance, l'étudiant sera en mesure d'effectuer des recherches de façon efficiente sur un ensemble de données.
+
+Le code présenté dans les énoncés se trouve sur Moodle, dans le dossier \lstinline{Ressources}.
+
+
+\section{Recherche binaire}
+
+% ADVANCED
+\begin{Exercice}[10 minutes] Recherche binaire (Python)\\
+
+    Soient une liste d’entiers triés \lstinline{L} et un entier \lstinline{e}. Écrivez un programme qui retourne l'index de l'élément \lstinline{e} de la liste L en utilisant une recherche binaire. Si \lstinline{e} n’est pas dans \lstinline{L}, retournez \lstinline{-1}.\\
+    
+    \begin{Example}{\faTerminal \quad Exemple}
+        L = [123, 321, 328, 472, 549] \\
+        e = 328\\
+    
+    Résultat attendu: 2
+    \end{Example}
+    
+    \lstinputlisting{resources/question1_advanced.py}
+    
+    \begin{conseil}
+        Inspirez-vous des conseils des exercices précédents (exercices basiques).
+    \end{conseil}
+    
+        \begin{solution}
+            \textbf{Python :}
+            \lstinputlisting{solutions/question1_advanced.py}
+        \end{solution}
+    
+    \end{Exercice}
+    
+
+% ADVANCED
+\begin{Exercice}[15 minutes] Recherche binaire - plus proche élément (Python)\\
+
+Soient une liste d’entiers \textbf{triés} \lstinline{L} et un entier \lstinline{e}. Écrivez un programme retournant la valeur dans \lstinline{L} la plus proche de \lstinline{e} en utilisant une recherche binaire (binary search).\\
+
+\lstinputlisting{resources/question2_advanced.py}
```

Comment:
La solution de cet exo ne marche pas dans certain cas. Si je donne un élément qui n'appartient pas à la liste il retourne une erreur (IndexError: list index out of range) et j'ai aussi testé avec la valeur 61 ca me fait une erreur.
J'inspecterai ça plus tard pendant la journée, sinon est ce que vous pouvez jeter un coup d'oeil @abiola123 et @LouayN.